firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -0,0 +1,198 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.remote;
+
+import android.content.Context;
+import android.support.annotation.VisibleForTesting;
+import com.google.android.gms.common.GooglePlayServicesNotAvailableException;
+import com.google.android.gms.common.GooglePlayServicesRepairableException;
+import com.google.android.gms.security.ProviderInstaller;
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.TaskCompletionSource;
+import com.google.android.gms.tasks.Tasks;
+import com.google.firebase.firestore.core.DatabaseInfo;
+import com.google.firebase.firestore.util.AsyncQueue;
+import com.google.firebase.firestore.util.Logger;
+import com.google.firebase.firestore.util.Supplier;
+import com.google.firestore.v1.FirestoreGrpc;
+import io.grpc.CallCredentials;
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.MethodDescriptor;
+import io.grpc.android.AndroidChannelBuilder;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Manages the gRPC channel and encapsulates all SSL and gRPC initialization.
+ *
+ * <p>All operations are dispatched to an internal worker queue and are not executed until the SSL
+ * and gRPC Stub initialization completes.
+ */
+// PORTING NOTE: This class only exists on Android.
+public class GrpcCallProvider {
+
+  private static final String LOG_TAG = "GrpcCallProvider";
+
+  private static Supplier<ManagedChannelBuilder<?>> overrideChannelBuilderSupplier;
+
+  private final Task<ManagedChannel> channelTask;
+  private final AsyncQueue asyncQueue;
+
+  private CallOptions callOptions;
+
+  /**
+   * Helper function to globally override the channel that RPCs use. Useful for testing when you
+   * want to bypass SSL certificate checking.
+   *
+   * @param channelBuilderSupplier The supplier for a channel builder that is used to create gRPC
+   *     channels.
+   */
+  @VisibleForTesting
+  public static void overrideChannelBuilder(
+      Supplier<ManagedChannelBuilder<?>> channelBuilderSupplier) {
+    overrideChannelBuilderSupplier = channelBuilderSupplier;
+  }
+
+  GrpcCallProvider(
+      AsyncQueue asyncQueue,
+      Context context,
+      DatabaseInfo databaseInfo,
+      CallCredentials firestoreHeaders) {
+    this.asyncQueue = asyncQueue;
+
+    TaskCompletionSource<ManagedChannel> channelReady = new TaskCompletionSource<>();
+    this.channelTask = channelReady.getTask();
+
+    // We execute network initialization on a separate thread to not block operations that
+    // depend on the AsyncQueue.
+    new Thread("FirestoreGrpcInit") {
```

Comment:
Rather than manually create our own thread, I wonder if it would be better to use `Tasks.call(AsyncTask.THREAD_POOL_EXECUTOR, ...)`

https://developers.google.com/android/reference/com/google/android/gms/tasks/Tasks.html#call(java.util.concurrent.Executor,%20java.util.concurrent.Callable%3CTResult%3E)
https://developer.android.com/reference/android/os/AsyncTask#THREAD_POOL_EXECUTOR

Unfortunately I don't see this pattern in existing Firebase code (or any usage of AsyncTask at all for that matter), so I'm a tiny bit scared there might be pitfalls.

Oh... Or even easier, can we just use https://developers.google.com/android/reference/com/google/android/gms/security/ProviderInstaller.html#installIfNeededAsync(android.content.Context,%20com.google.android.gms.security.ProviderInstaller.ProviderInstallListener) and not mess with threads ourselves?  That would call us back on the UI thread which isn't quite ideal but might be fine...

Reply:
Re 2: That works, but it doesn't allow me to move the Stub initialization. I discovered the Async method on Friday when I was already fully committed to the idea that we should also move the initialization of the GRPC stub.

Reply:
Heh.  Is the initialization of the gRPC stub measurably slow? :)

Reply:
In my perf findings doc, of the 80ms we were spending on initialization, only half of it was the ProviderInstaller. Another ~30ms was gRPC stubs.

Reply:
Cool, thanks.

Reply:
Changed the code to run on Android's magical AsyncTask queue. I feel like a proper Android developer now.